### ISSUE TITLE
Error calling pg_stat_statements_reset() as requested intermittently

### ISSUE THREAD

**Issue by dbamohsin** [pganalyze/collector#398]
Hi,

we have been investigating a perceived memory leak in our collector and have contributed to the previous issue that was raised (https://github.com/pganalyze/collector/issues/37) - on the back of that issue being closed we started looking into what is actually causing *some* of our collectors to randomly restart.

We see memory growing in a linear fashion until the pod restarts:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/61407098/225260446-fb14e2eb-2bb4-4b89-89dd-b7c54f5c3ab4.png">

The above screenshot shows 2 collectors, one that faces the issue, and another that doesnt.

We then tried to match up what happens at the start of memory jump, and we were able to see an error in the logs that matches up to the start point of the memory growth on each occasion..

```
2023-03-13T02:20:23.006696777Z E [default] Error calling pg_stat_statements_reset() as requested: pq: permission denied for function pg_stat_statements_reset
2023-03-13T02:20:23.006746962Z E [default] Could not process server: pq: permission denied for function pg_stat_statements_reset
```

From what we can see, some collectors seem to think that they dont have permissions on the database and then are unable to send stats - which is strange as its all very intermittent. once the collector crashes and restarts - it all starts working again.

Once we get the error, we never see a `Submitted snapshot successfully` until the collector restarts (when it hits its memory limit). So our theory is that once this error happens, the collector is building up a snapshot locally and thus the memory grows, and as it isnt able to send it to pganalyze, we eventually hit the limit of what it can store in memory locally on the pod.

We'd like to get to the bottom of what causes it to lose its permissions as we think this is the root cause.

Thanks

**Comment by dbamohsin** [comment]
So ive noticed that on stock-pg (the collector that has the issue) we had Collector: Query Statistics Reset Frequency set - so i reckon this is something that we have done historically.. ive just disabled that to see impact

**Comment by lfittl** [comment]
@dbamohsin Thanks for the report!

Just as a quick initial note, the error message you posted is directly related to the `Collector: Query Statistics Reset Frequency` - the collector only calls pg_stat_statements_reset when that setting is enabled. In order to make use of this setting, you typically need to have the following helper function in place on the database:

```sql
CREATE OR REPLACE FUNCTION pganalyze.reset_stat_statements() RETURNS SETOF void AS
$$
  /* pganalyze-collector */ SELECT * FROM public.pg_stat_statements_reset();
$$ LANGUAGE sql VOLATILE SECURITY DEFINER;
```

Separately, we'll take a look to see if we can confirm that the reset erroring out causes a memory leak (it could be that the error state isn't cleaned up correctly somehow) - the data you provided should be an excellent starting point to pinpoint the problem.
